### PR TITLE
test: use snapshot

### PR DESCRIPTION
**Pull request by marcalexiei**
Part of #141

---

### Snapshots

Snapshots are used whenever is possible.

> [!WARNING]
> It was not possible to use them in all tests.

<details>
<summary>Detailed explanation</summary>

#### `watchlist`

Can't use snapshot since is comparing rollup properties with file system files. 
However I was able to simplify the logic (I created a separate commit).

#### `output`

Using snapshot here is not feasible due to rollup:
It looks like that, randomly, plugin `transform` is called first on `actual_b` and then on `actual_a` instead of calling transform on  `actual_a` and then on `actual_b`.

So instead of having:
```
body{color:red}body{color:green}
```
you end up having
```
body{color:green}body{color:red}
```
But only some times.

This looks like a rollup bug:
- https://github.com/rollup/rollup/issues/3888

Right now on `main` branch `should support output as function` sometimes fails for the exact same reason.
Also on other tests involving `output`, assertions are written using `indexOf !== -1` logic.
I assume it's because of this bug.

E.g.:
https://github.com/elycruz/rollup-plugin-sass/blob/0425739612e38ded0fda00e5060b25a735d620c3/test/index.test.ts#L265-L268

Right now there isn't much we can do. 
I'm going to add a code comment on `options.output.test.ts` referencing this PR

</details>

### `insertStyle.test.ts` – replace `jsdom` with `happy-dom`

Since the test are quite simple I opted out to switch `jsdom` to `happy-dom` (less dependencies and faster)

## Additional changes

- All unit test dependencies have been updated:
  -  `ava`
  - `@ava/typescript`
  - `nyc`
  - `sinon`
- Added missing `@types` for
  -  `sinon`
  -  `icss-utils`

### CODE REVIEW

**Comment by marcalexiei** [comment]
All review comments of #143 have been applied in this PR

File: test/insertStyle.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  global["window"] = page.mainFrame.window;
+  // @ts-expect-error
+  global["document"] = page.mainFrame.window.document;
+
+  const cssStr = insertStyle(expectA);
+
+  const styleSheet = document.head.querySelector("style")!;
+  t.is(
+    styleSheet.textContent,
+    cssStr!,
```
Comment (elycruz) [inline test/insertStyle.test.ts:24]: Erroneous `!` char.

File: test/insertStyle.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "text/css",
+    'Should contain `type` attrib. equal to "text/css"'
+  );
+
+  await browser.close();
+});
+
+test("insertStyle shouldn't choke when window is undefined", (t) => {
+  delete global["window"];
+  t.throws(() => !window);
```
Comment (elycruz) [inline test/insertStyle.test.ts:38]: `throw` test seems un-required here - reasoning?

File: test/snapshots/test/index.test.ts.snap
Comment (elycruz) [inline test/snapshots/test/index.test.ts.snap:1]: Do we need to track these *.snap files?

**Review by elycruz — COMMENTED** [review]
Left a couple of comments - Review of test/index.test.ts still pending.

File: test/insertStyle.test.ts (hunk 1 above)
Comment (marcalexiei) [reply test/insertStyle.test.ts:24]: `cssStr` is `string | undefined` and `t.is` accepts `string | null` so the not null assertion is required to remove `undefined` from `cssStr`  to make it assignable to `t.is` second parameter type.

image

File: test/insertStyle.test.ts (hunk 2 above)
Comment (marcalexiei) [reply test/insertStyle.test.ts:38]: I can only guess the reason... this assertion is already present on the main branch
https://github.com/elycruz/rollup-plugin-sass/blob/0425739612e38ded0fda00e5060b25a735d620c3/test/style.test.ts#L31

the tests are probably executed in `strict` mode so an error occurs if you access a variable that doesn't exists

File: test/snapshots/test/index.test.ts.snap
Comment (marcalexiei) [reply test/snapshots/test/index.test.ts.snap:1]: https://github.com/avajs/ava/blob/main/docs/04-snapshot-testing.md

> Say you have `~/project/test/main.js` which contains snapshot assertions. AVA will create two files:
>
> * `~/project/test/snapshots/main.js.snap`
> * `~/project/test/snapshots/main.js.md`
>
> The first file contains the actual snapshot and is required for future comparisons.
> The second file contains your *snapshot report*. It's regenerated when you update your snapshots. 
> If you commit it to source control you can diff it to see the changes to your snapshot.

File: test/insertStyle.test.ts (hunk 1 above)
Comment (marcalexiei) [reply test/insertStyle.test.ts:24]: If you prefer I can switch to null coalescing:
```suggestion
    cssStr ?? null,
```

**Comment by marcalexiei** [comment]
@elycruz yesterday you made changes to test on main via #153 leading to conflicts with the work I done here.
Snapshot test update PRs has been opened for nearly 2 months (#143) so could you please complete the review?
If you changed your mind feel free to close this.

**Comment by elycruz** [comment]
@marcalexiei Hey, right - apologies about that - have been uber swamped lately -
Not a problem - I'm merging it into the 'dev' branch (will make some annotations) and will add you to the subsequent PR - 
I'll correct/make-small-updates for annotations I'm making and then after you review those changes we can merge it.

**Pull request merged**